### PR TITLE
Fix for Powermock integration

### PR DESCRIPTION
**Pull request by karlicoss**
Fixes the issue I raised earlier: https://github.com/robolectric/robolectric/issues/2208 and will hopefully fix https://github.com/robolectric/robolectric/issues/2153. Basically, the tests using Powermock were failing with

		com.thoughtworks.xstream.converters.ConversionException: Cannot convert type org.robolectric.internal.dependency.CachedDependencyResolver to type org.robolectric.internal.dependency.DependencyResolver
		---- Debugging information ----
		class               : org.robolectric.MultiApiRobolectricTestRunner$TestRunnerForApiVersion
		required-type       : org.robolectric.MultiApiRobolectricTestRunner$TestRunnerForApiVersion
		converter-type      : com.thoughtworks.xstream.converters.reflection.ReflectionConverter
		path                : /org.powermock.modules.junit4.rule.PowerMockStatement$1/outer-class/fNext/this$1/outer-class/dependencyResolver
		line number         : 250
		class[1]            : org.robolectric.MultiApiRobolectricTestRunner$TestRunnerForApiVersion$1
		class[2]            : org.robolectric.RobolectricTestRunner$HelperTestRunner$1
		class[3]            : org.powermock.modules.junit4.rule.PowerMockStatement
		class[4]            : org.powermock.modules.junit4.rule.PowerMockStatement$1
		version             : not available
		-------------------------------

			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:439)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
			at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
			at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
			at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
			at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:480)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:412)
			at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:263)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
			at com.thoughtworks.xstream.core.AbstractReferenceUnmarshaller.convert(AbstractReferenceUnmarshaller.java:65)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
			at com.thoughtworks.xstream.core.TreeUnmarshaller.start(TreeUnmarshaller.java:134)
			at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.unmarshal(AbstractTreeMarshallingStrategy.java:32)
			at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1206)
			at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1190)
			at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1061)
			at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1052)
			at org.powermock.classloading.DeepCloner.clone(DeepCloner.java:54)
			at org.powermock.classloading.ClassloaderExecutor.execute(ClassloaderExecutor.java:89)
			at org.powermock.classloading.ClassloaderExecutor.execute(ClassloaderExecutor.java:78)
			at org.powermock.modules.junit4.rule.PowerMockStatement.evaluate(PowerMockRule.java:57)
			at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:255)
			at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:185)
			at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
			at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
			at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
			at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
			at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
			at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
			at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:151)
			at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
			at org.junit.runners.Suite.runChild(Suite.java:128)
			at org.junit.runners.Suite.runChild(Suite.java:24)
			at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
			at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
			at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
			at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
			at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
			at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
			at org.junit.runner.JUnitCore.run(JUnitCore.java:157)
			at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
			at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
			at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
			at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
			at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
			at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
			at java.lang.reflect.Method.invoke(Method.java:497)
			at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)

I feel like the PR is unfinished anyway since I left some mess in the `pom.xml` file which has to be fixed, and would be nice if someone clarified some internal stuff, so please respond to the comments :)

### CODE REVIEW

File: robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java [hunk 1]
```diff
@@ -152,6 +153,7 @@ public InstrumentationConfiguration build() {
           DoNotInstrument.class,
           Config.class,
           Transcript.class,
+          DependencyResolver.class,
```
Comment (karlicoss) [inline robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java]: Apparently, there was some mess with classloading, and `InstrumentingClassLoader` tried instrumenting `CachedDependencyResolver`, which caused classloaders mismatch and the exception above.

Actually, I am somewhat surprised to see (at the first sight, arbitrary) internal Robolectric classes listed here. I don't have a deep understanding of the `InstrumentingClassLoader`, but shouldn't it only load android-specific classes? Is there really a need to explicitly specify Robolectric classes to exclude? Or am I missing something?

File: robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java [hunk 2]
```diff
@@ -170,7 +172,8 @@ public InstrumentationConfiguration build() {
           "org.specs2",  // allows for android projects with mixed scala\java tests to be
           "scala.",      //  run with Maven Surefire (see the RoboSpecs project on github)
           "kotlin.",
-          "com.almworks.sqlite4java" // Fix #958: SQLite native library must be loaded once.
+          "com.almworks.sqlite4java", // Fix #958: SQLite native library must be loaded once.
+          "org.powermock"
```
Comment (karlicoss) [inline robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java]: without this, I get

	java.lang.IllegalAccessError: tried to access field org.robolectric.RobolectricTestRunner$HelperTestRunner.this$0 from class org.robolectric.RobolectricTestRunner$HelperTestRunner$1

		at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:545)
		at org.powermock.modules.junit4.rule.PowerMockStatement$1.run(PowerMockRule.java:65)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:497)
		at org.powermock.reflect.internal.WhiteboxImpl.performMethodInvocation(WhiteboxImpl.java:1873)
		at org.powermock.reflect.internal.WhiteboxImpl.doInvokeMethod(WhiteboxImpl.java:773)
		at org.powermock.reflect.internal.WhiteboxImpl.invokeMethod(WhiteboxImpl.java:638)
		at org.powermock.reflect.Whitebox.invokeMethod(Whitebox.java:401)
		at org.powermock.classloading.ClassloaderExecutor.execute(ClassloaderExecutor.java:98)
		at org.powermock.classloading.ClassloaderExecutor.execute(ClassloaderExecutor.java:78)
		at org.powermock.modules.junit4.rule.PowerMockStatement.evaluate(PowerMockRule.java:57)
		at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:255)
		at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:185)
		at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
		at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
		at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
		at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
		at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
		at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
		at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:151)
       ...

Even though `packagesToNotAquire` is configurable externally, I guess it makes sense to include powermock by default, it wouldn't hurt, right?

**Comment by karlicoss** [comment]
oops, got test failures, closing until fix them

**Comment by hzsweers** [comment]
@karlicoss were you able to get anywhere with this?

**Comment by karlicoss** [comment]
@hzsweers I digged into it, but got distracted (for quite a while, apparently). I will probably resume at some point.

**Comment by karlicoss** [comment]
@hzsweers actually, some instrumentation refactoring was done recently, so I will check, maybe it got fixed 'by accident' :D

**Comment by hzsweers** [comment]
Unless it was in the last 24 hours, it's still broken :/

**Comment by karlicoss** [comment]
Ah okay. I'll check it anyway, maybe I'll at least post results of my investigation

**Comment by hzsweers** [comment]
Thanks, looking forward to it!

**Comment by khayamy001** [comment]
I have the same problem, thanks for taking the time to look into it :)

**Comment by karlicoss** [comment]
Hey guys, sorry for the delay, I'm just way too busy with my work and my
master thesis stuff right now. It also turns out I need this resolved as
well, so I will continue looking at it soon :)

On Thu, Mar 17, 2016 at 3:21 PM, Youssef Elkhayami <notifications@github.com
> wrote:

> I have the same problem, thanks for taking the time to look into it :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/robolectric/robolectric/pull/2213#issuecomment-197854558>
>



-- 
Dmitrii Gerasimov, ITMO University, Department of Higher Mathematics, study
group A4200

File: robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java (hunk 1 above)
Comment (jongerrish) [inline robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java]: I noticed that also and think that it is broken. Seems there are some types loaded by the system classsloader that shouldn't be. I believe all Android types should be loaded by the instrumenting classloader.

File: robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java (hunk 2 above)
Comment (jongerrish) [inline robolectric/src/main/java/org/robolectric/internal/bytecode/InstrumentationConfiguration.java]: I think that packagesToNotAquire should just be simplified as all non-android classes, but again I think there is a leaky / undefined boundary by what is loaded by the instrumenting class loader. I think this needs fixing.

**Comment by hzsweers** [comment]
@jongerrish any thoughts on what next steps here should be? I'm not totally familiar with robolectric internals, but happy to help any way I can to get this resolved so we can start using snapshots.

**Pull request closed without merging**